subsys/usb/device_next/class/usbd_hid.c

```diff
@@ -0,0 +1,743 @@
+		return 0;
+	}
+
+	if (desc->if0.bInterfaceSubClass == 0) {
+		/* The device does not support the boot protocol */
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	LOG_DBG("Get Protocol: %s", ddata->protocol ? "Report" : "Boot");
+	net_buf_add_u8(buf, ddata->protocol);
+
+	return 0;
+}
+
+static int handle_get_descriptor(const struct device *dev,
+				 const struct usb_setup_packet *const setup,
+				 struct net_buf *const buf)
+{
+	struct hid_device_data *const ddata = dev->data;
+	uint8_t desc_type = USB_GET_DESCRIPTOR_TYPE(setup->wValue);
+	uint8_t desc_idx = USB_GET_DESCRIPTOR_INDEX(setup->wValue);
+	struct usbd_hid_descriptor *const desc = ddata->desc;
+
+	switch (desc_type) {
+	case USB_DESC_HID_REPORT:
+		LOG_DBG("Get descriptor report");
+		net_buf_add_mem(buf, ddata->rdesc, MIN(ddata->rsize, setup->wLength));
+		break;
+	case USB_DESC_HID:
+		LOG_DBG("Get descriptor HID");
+		net_buf_add_mem(buf, &desc->hid, MIN(desc->hid.bLength, setup->wLength));
+		break;
+	case USB_DESC_HID_PHYSICAL:
+		LOG_DBG("Get descriptor physical %u", desc_idx);
+		errno = -ENOTSUP;
+		break;
+	default:
+		errno = -ENOTSUP;
+		break;
+	}
+
+	return 0;
+}
+
+static int usbd_hid_ctd(struct usbd_class_data *const c_data,
+			const struct usb_setup_packet *const setup,
+			const struct net_buf *const buf)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	int ret = 0;
+
+	switch (setup->bRequest) {
+	case USB_HID_SET_IDLE:
+		ret = handle_set_idle(dev, setup);
+		break;
+	case USB_HID_SET_REPORT:
+		ret = handle_set_report(dev, setup, buf);
+		break;
+	case USB_HID_SET_PROTOCOL:
+		ret = handle_set_protocol(dev, setup);
+		break;
+	default:
+		errno = -ENOTSUP;
+		break;
+	}
+
+	return ret;
+}
+
+static int usbd_hid_cth(struct usbd_class_data *const c_data,
+			const struct usb_setup_packet *const setup,
+			struct net_buf *const buf)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	int ret = 0;
+
+	switch (setup->bRequest) {
+	case USB_HID_GET_IDLE:
+		ret = handle_get_idle(dev, setup, buf);
+		break;
+	case USB_HID_GET_REPORT:
+		ret = handle_get_report(dev, setup, buf);
+		break;
+	case USB_HID_GET_PROTOCOL:
+		ret = handle_get_protocol(dev, setup, buf);
+		break;
+	case USB_SREQ_GET_DESCRIPTOR:
+		ret = handle_get_descriptor(dev, setup, buf);
+		break;
+	default:
+		errno = -ENOTSUP;
+		break;
+	}
+
+	return ret;
+}
+
+static void usbd_hid_sof(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct hid_device_data *ddata = dev->data;
+	const struct hid_device_ops *const ops = ddata->ops;
+
+	if (ops->sof) {
+		ops->sof(dev);
+	}
+}
+
+static void usbd_hid_enable(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct hid_device_data *ddata = dev->data;
+	const struct hid_device_ops *const ops = ddata->ops;
+	struct usbd_hid_descriptor *const desc = ddata->desc;
+
+	atomic_set_bit(&ddata->state, HID_DEV_CLASS_ENABLED);
```

Comment:
I do not understand what you are getting at and it is hard to guess. What are your current concerns?

Reply:
E.g. `hid_dev_submit_report` may be called by an application from any context (also preemptive). Then it is possible that function execution would be preempted after `HID_DEV_CLASS_ENABLED` flag check and `usbd_ep_enqueue` would be called after HID device class was actually disabled. Is it acceptable?

Reply:
Also the `hid_dev_output_handler` may be preempted in similar way (if someone uses preemptive system workqueue priority)

Reply:
> Then it is possible that function execution would be preempted after `HID_DEV_CLASS_ENABLED` flag check and `usbd_ep_enqueue` would be called after HID device class was actually disabled. Is it acceptable?

Yes, it will fail and return an error.

Reply:
Ok, so an application must handle this case if needed.